File: src-tauri/src/ws_server/dispatch_fs.rs
//! Filesystem command dispatch for the WebSocket server.
//!
//! Handles all `fs_*` commands plus filesystem-related commands from other modules.

use super::dispatch_helpers::extract_arg;
use super::WsState;

/// Dispatch a filesystem command, returning the JSON result.
pub async fn dispatch(
    cmd: &str,
    args: serde_json::Value,
    _state: &WsState,
) -> Result<serde_json::Value, String> {
    match cmd {
        "fs_read_file" => {
            let path: String = extract_arg(&args, "path")?;
            let result = crate::filesystem::read_file(&path)?;
            Ok(serde_json::to_value(result).unwrap())
        }
        "fs_exists" => {
            let path: String = extract_arg(&args, "path")?;
            let result = crate::filesystem::exists(&path);
            Ok(serde_json::to_value(result).unwrap())
        }
        "fs_list_dir" => {
            let path: String = extract_arg(&args, "path")?;
            let result = crate::filesystem::list_dir(&path)?;
            Ok(serde_json::to_value(result).unwrap())
        }
        "fs_write_file" => {
            let path: String = extract_arg(&args, "path")?;
            let contents: String = extract_arg(&args, "contents")?;
            crate::filesystem::fs_write_file(path, contents)?;
            Ok(serde_json::Value::Null)
        }
        "fs_mkdir" => {
            let path: String = extract_arg(&args, "path")?;
            crate::filesystem::fs_mkdir(path)?;
            Ok(serde_json::Value::Null)
        }
        "fs_remove" => {
            let path: String = extract_arg(&args, "path")?;
            crate::filesystem::fs_remove(path)?;
            Ok(serde_json::Value::Null)
        }
        "fs_remove_dir_all" => {
            let path: String = extract_arg(&args, "path")?;
            crate::filesystem::fs_remove_dir_all(path)?;
            Ok(serde_json::Value::Null)
        }
        "fs_move" => {
            let from: String = extract_arg(&args, "from")?;
            let to: String = extract_arg(&args, "to")?;
            crate::filesystem::fs_move(from, to)?;
            Ok(serde_json::Value::Null)
        }
        "fs_copy_file" => {
            let from: String = extract_arg(&args, "from")?;
            let to: String = extract_arg(&args, "to")?;
            crate::filesystem::fs_copy_file(from, to)?;
            Ok(serde_json::Value::Null)
        }
        "fs_copy_directory" => {
            let from: String = extract_arg(&args, "from")?;
            let to: String = extract_arg(&args, "to")?;
            crate::filesystem::fs_copy_directory(from, to)?;
            Ok(serde_json::Value::Null)
        }
        "fs_is_git_repo" => {
            let path: String = extract_arg(&args, "path")?;
            let result = crate::filesystem::fs_is_git_repo(path);
            Ok(serde_json::to_value(result).unwrap())
        }
        "fs_git_worktree_add" => {
            let repo_path: String = extract_arg(&args, "repoPath")?;
            let worktree_path: String = extract_arg(&args, "worktreePath")?;
            crate::filesystem::fs_git_worktree_add(repo_path, worktree_path)?;
            Ok(serde_json::Value::Null)
        }
        "fs_git_worktree_remove" => {
            let repo_path: String = extract_arg(&args, "repoPath")?;
            let worktree_path: String = extract_arg(&args, "worktreePath")?;
            crate::filesystem::fs_git_worktree_remove(repo_path, worktree_path)?;
            Ok(serde_json::Value::Null)
        }
        "fs_grep" => {
            let dir: String = extract_arg(&args, "dir")?;
            let pattern: String = extract_arg(&args, "pattern")?;
            let file_glob: String = extract_arg(&args, "fileGlob")?;
            let result = crate::filesystem::grep(&dir, &pattern, &file_glob)?;
            Ok(serde_json::to_value(result).unwrap())
        }
        "fs_bulk_read" => {
            let paths: Vec<String> = serde_json::from_value(
                args.get("paths").cloned().ok_or("Missing 'paths'")?
            ).map_err(|e| format!("Invalid paths: {}", e))?;
            let result = crate::filesystem::bulk_read(&paths);
            Ok(serde_json::to_value(result).unwrap())
        }
        // Mort filesystem commands
        "fs_get_repo_dir" => {
            let repo_name: String = extract_arg(&args, "repoName")?;
            let result = crate::mort_commands::fs_get_repo_dir(repo_name).await?;
            Ok(serde_json::to_value(result).unwrap())
        }
        "fs_get_repo_source_path" => {
            let repo_name: String = extract_arg(&args, "repoName")?;
            let result = crate::mort_commands::fs_get_repo_source_path(repo_name).await?;
            Ok(serde_json::to_value(result).unwrap())
        }
        "fs_get_home_dir" => {
            let result = crate::mort_commands::fs_get_home_dir().await?;
            Ok(serde_json::to_value(result).unwrap())
        }
        "fs_list_dir_names" => {
            let path: String = extract_arg(&args, "path")?;
            let result = crate::mort_commands::fs_list_dir_names(path).await?;
            Ok(serde_json::to_value(result).unwrap())
        }
        _ => Err(format!("unknown fs command: {}", cmd)),
    }
}

File: sidecar/src/dispatch/dispatch-fs.ts
/**
 * Filesystem command dispatch.
 * Handles all `fs_*` commands.
 */

import {
  readFile,
  writeFile,
  mkdir,
  rm,
  rename,
  stat,
  readdir,
  cp,
  copyFile,
} from "node:fs/promises";
import { existsSync } from "node:fs";
import { execFile } from "node:child_process";
import { promisify } from "node:util";
import { join } from "node:path";
import { extractArg } from "../helpers.js";
import { dataDirPath, homeDirPath } from "./paths.js";
import type { SidecarState } from "../state.js";

const execFileAsync = promisify(execFile);

interface DirEntry {
  name: string;
  path: string;
  isDirectory: boolean;
  isFile: boolean;
}

export async function dispatchFs(
  cmd: string,
  args: Record<string, unknown>,
  _state: SidecarState,
): Promise<unknown> {
  switch (cmd) {
    case "fs_read_file":
      return readFile(extractArg<string>(args, "path"), "utf-8");

    case "fs_exists":
      return existsSync(extractArg<string>(args, "path"));

    case "fs_list_dir":
      return listDir(extractArg<string>(args, "path"));

    case "fs_write_file": {
      const path = extractArg<string>(args, "path");
      await mkdir(join(path, ".."), { recursive: true });
      await writeFile(path, extractArg<string>(args, "contents"));
      return null;
    }

    case "fs_mkdir":
      await mkdir(extractArg<string>(args, "path"), { recursive: true });
      return null;

    case "fs_remove":
      await rm(extractArg<string>(args, "path"), { force: true });
      return null;

    case "fs_remove_dir_all":
      await rm(extractArg<string>(args, "path"), {
        recursive: true,
        force: true,
      });
      return null;

    case "fs_move":
      await rename(
        extractArg<string>(args, "from"),
        extractArg<string>(args, "to"),
      );
      return null;

    case "fs_copy_file":
      await copyFile(
        extractArg<string>(args, "from"),
        extractArg<string>(args, "to"),
      );
      return null;

    case "fs_copy_directory":
      await cp(
        extractArg<string>(args, "from"),
        extractArg<string>(args, "to"),
        { recursive: true },
      );
      return null;

    case "fs_is_git_repo":
      return isGitRepo(extractArg<string>(args, "path"));

    case "fs_git_worktree_add":
      return fsGitWorktreeAdd(
        extractArg<string>(args, "repoPath"),
        extractArg<string>(args, "worktreePath"),
      );

    case "fs_git_worktree_remove":
      return fsGitWorktreeRemove(
        extractArg<string>(args, "repoPath"),
        extractArg<string>(args, "worktreePath"),
      );

    case "fs_grep":
      return fsGrep(
        extractArg<string>(args, "dir"),
        extractArg<string>(args, "pattern"),
        extractArg<string>(args, "fileGlob"),
      );

    case "fs_write_binary": {
      const path = extractArg<string>(args, "path");
      const data = Buffer.from(
        extractArg<string>(args, "base64Data"),
        "base64",
      );
      await writeFile(path, data);
      return null;
    }

    case "fs_bulk_read":
      return bulkRead(extractArg<string[]>(args, "paths"));

    case "fs_get_repo_dir": {
      const repoName = extractArg<string>(args, "repoName");
      return join(dataDirPath(), "repositories", repoName);
    }

    case "fs_get_repo_source_path":
      return getRepoSourcePath(extractArg<string>(args, "repoName"));

    case "fs_get_home_dir":
      return homeDirPath();

    case "fs_list_dir_names":
      return listDirNames(extractArg<string>(args, "path"));

    default:
      throw new Error(`unknown fs command: ${cmd}`);
  }
}

async function listDir(path: string): Promise<DirEntry[]> {
  const entries = await readdir(path, { withFileTypes: true });
  return entries.map((e) => ({
    name: e.name,
    path: join(path, e.name),
    isDirectory: e.isDirectory(),
    isFile: e.isFile(),
  }));
}

function isGitRepo(path: string): boolean {
  return existsSync(join(path, ".git")) || existsSync(join(path, "HEAD"));
}

async function fsGitWorktreeAdd(
  repoPath: string,
  worktreePath: string,
): Promise<null> {
  await execFileAsync("git", ["worktree", "prune"], { cwd: repoPath });
  await execFileAsync(
    "git",
    ["worktree", "add", "--detach", "--force", worktreePath, "HEAD"],
    { cwd: repoPath },
  );
  return null;
}

async function fsGitWorktreeRemove(
  repoPath: string,
  worktreePath: string,
): Promise<null> {
  await execFileAsync(
    "git",
    ["worktree", "remove", "--force", worktreePath],
    { cwd: repoPath },
  );
  return null;
}

async function fsGrep(
  dir: string,
  pattern: string,
  fileGlob: string,
): Promise<{ path: string; line: string; lineNumber: number }[]> {
  const results: { path: string; line: string; lineNumber: number }[] = [];
  const regex = new RegExp(pattern);

  const entries = await readdir(dir, { withFileTypes: true });
  for (const entry of entries) {
    if (!entry.isDirectory()) continue;
    const filePath = join(dir, entry.name, fileGlob);
    try {
      const content = await readFile(filePath, "utf-8");
      const lines = content.split("\n");
      for (let i = 0; i < lines.length; i++) {
        if (regex.test(lines[i])) {
          results.push({ path: filePath, line: lines[i], lineNumber: i + 1 });
        }
      }
    } catch {
      // File doesn't exist in this subdirectory — skip
    }
  }
  return results;
}

async function bulkRead(paths: string[]): Promise<(string | null)[]> {
  return Promise.all(
    paths.map(async (p) => {
      try {
        return await readFile(p, "utf-8");
      } catch {
        return null;
      }
    }),
  );
}

async function getRepoSourcePath(repoName: string): Promise<string> {
  const repoDir = join(dataDirPath(), "repositories", repoName);
  try {
    const settings = JSON.parse(
      await readFile(join(repoDir, "settings.json"), "utf-8"),
    );
    if (settings.sourcePath) return settings.sourcePath;
  } catch {
    // Try metadata.json fallback
  }
  try {
    const meta = JSON.parse(
      await readFile(join(repoDir, "metadata.json"), "utf-8"),
    );
    if (meta.sourcePath) return meta.sourcePath;
  } catch {
    // Neither file found
  }
  throw new Error(`Could not find sourcePath for repo: ${repoName}`);
}

async function listDirNames(path: string): Promise<string[]> {
  const entries = await readdir(path, { withFileTypes: true });
  return entries.map((e) => e.name);
}
